exporter/awsxrayexporter/translator/segment.go

```diff
@@ -0,0 +1,320 @@
+	Cause       *CauseData `json:"cause,omitempty"`
+	ResourceARN string     `json:"resource_arn,omitempty"`
+	Origin      string     `json:"origin,omitempty"`
+
+	Type         string   `json:"type,omitempty"`
+	Namespace    string   `json:"namespace,omitempty"`
+	User         string   `json:"user,omitempty"`
+	PrecursorIDs []string `json:"precursor_ids,omitempty"`
+
+	HTTP *HTTPData `json:"http,omitempty"`
+	AWS  *AWSData  `json:"aws,omitempty"`
+
+	Service *ServiceData `json:"service,omitempty"`
+
+	// SQL
+	SQL *SQLData `json:"sql,omitempty"`
+
+	// Metadata
+	Annotations map[string]interface{}            `json:"annotations,omitempty"`
+	Metadata    map[string]map[string]interface{} `json:"metadata,omitempty"`
+}
+
+var (
+	writers = newWriterPool(2048)
+)
+
+// MakeSegmentDocumentString converts an OpenCensus Span to an X-Ray Segment and then serialzies to JSON
+func MakeSegmentDocumentString(name string, span *tracepb.Span) (string, error) {
+	segment := MakeSegment(name, span)
+	w := writers.borrow()
+	if err := w.Encode(segment); err != nil {
+		return "", err
+	}
+	jsonStr := w.String()
+	writers.release(w)
+	return jsonStr, nil
+}
+
+// MakeSegment converts an OpenCensus Span to an X-Ray Segment
+func MakeSegment(name string, span *tracepb.Span) Segment {
+	var (
+		traceID                                = convertToAmazonTraceID(span.TraceId)
+		startTime                              = timestampToFloatSeconds(span.StartTime, span.StartTime)
+		endTime                                = timestampToFloatSeconds(span.EndTime, span.StartTime)
+		httpfiltered, http                     = makeHTTP(span)
+		isError, isFault, causefiltered, cause = makeCause(span.Status, httpfiltered)
+		isThrottled                            = span.Status.Code == tracetranslator.OCResourceExhausted
+		origin                                 = determineAwsOrigin(span.Resource)
+		awsfiltered, aws                       = makeAws(causefiltered, span.Resource)
+		service                                = makeService(span.Resource)
+		sqlfiltered, sql                       = makeSQL(awsfiltered)
+		user, annotations                      = makeAnnotations(sqlfiltered)
+		namespace                              string
+	)
+
+	if name == "" {
+		name = fixSegmentName(span.Name.String())
+	}
+	if span.ParentSpanId != nil {
+		namespace = "remote"
+	}
+
+	return Segment{
+		ID:          convertToAmazonSpanID(span.SpanId),
+		TraceID:     traceID,
+		Name:        name,
+		StartTime:   startTime,
+		EndTime:     endTime,
+		ParentID:    convertToAmazonSpanID(span.ParentSpanId),
+		Fault:       isFault,
+		Error:       isError,
+		Throttle:    isThrottled,
+		Cause:       cause,
+		Origin:      origin,
+		Namespace:   namespace,
+		User:        user,
+		HTTP:        http,
+		AWS:         aws,
+		Service:     service,
+		SQL:         sql,
+		Annotations: annotations,
+		Metadata:    nil,
+	}
+}
+
+// NewTraceID generates a new valid X-Ray TraceID
+func NewTraceID() []byte {
```

Comment:
This is only used in test. Does this need to be exported?

Reply:
Fixed